src/ooni/web_connectivity.cpp

```diff
@@ -0,0 +1,677 @@
+      (*entry)["blocking"] = "dns";
+    }
+    if ((*entry)["blocking"] != nullptr) {
+      std::string blocking = (*entry)["blocking"];
+      logger->info("web_connectivity: BLOCKING detected due to: %s on %s",
+                   blocking.c_str(), input.c_str());
+    } else {
+      logger->info("web_connectivity: no blocking detected");
+    }
+}
+
+static void control_request(Var<Entry> entry,
+        SocketList socket_list, std::string url,
+        Callback<Error> callback,
+        Settings options, Var<Logger> logger) {
+    Settings request_settings;
+    http::Headers headers;
+    Entry request;
+    request["tcp_connect"] = Entry::array();
+    for (auto socket : socket_list) {
+      // Formats the sockets as IP:PORT
+      std::ostringstream ss;
+      if (is_private_ipv4_addr(socket.first) == true) {
+         continue;
+      }
+      ss << socket.first;
+      ss << ":";
+      ss << socket.second;
+      request["tcp_connect"].push_back(ss.str());
+    }
+    request["http_request"] = url;
+    std::string body = request.dump();
+
+    request_settings["http/url"] = options["backend"];
+    request_settings["http/method"] = "POST";
+    headers["Content-Type"] = "application/json";
+
+    if (options["backend/type"] == "cloudfront") {
+      // TODO set the appropriate headers to support cloud-fronting.
+    }
+
+    logger->debug("web_connectivity: performing control request to %s",
+                  options["backend"].c_str());
+    http::request(request_settings, headers, body,
+            [=](Error error, Var<http::Response> response) {
+      if (!error) {
+        try {
+          (*entry)["control"] = Entry::parse(response->body);
+          callback(NoError());
+          return;
+        } catch (const std::invalid_argument &) {
+          (*entry)["control_failure"] = "json_parse_error";
+          callback(JsonParseError());
+          return;
+        }
+      }
+      (*entry)["control_failure"] = error.as_ooni_error();
+      callback(error);
+      return;
+    });
+}
+
+static void experiment_http_request(Var<Entry> entry,
+        std::string url, Callback<Error> cb,
+        Settings options,
+        Var<Reactor> reactor, Var<Logger> logger) {
+
+  http::Headers headers = COMMON_CLIENT_HEADERS;
+  std::string body;
+  options["http/url"] = url;
+
+  logger->debug("Requesting url %s", url.c_str());
+  templates::http_request(entry, options, headers, body, [=](Error err,
+              Var<http::Response> response) {
+    if (err) {
+      (*entry)["http_experiment_failure"] = err.as_ooni_error();
+      cb(err);
+      return;
+    }
+    cb(NoError());
+  }, reactor, logger);
+}
+
+static void experiment_tcp_connect(Var<Entry> entry,
+        SocketList sockets, Callback<Error> cb,
+        Var<Reactor> reactor, Var<Logger> logger) {
+
+  int socket_count = sockets.size();
+  Var<int> sockets_tested(new int(0));
+  // XXX this is very ghetto
+  if (socket_count == 0) {
+    cb(NoError());
```

Comment:
It actually doesn't matter since the other things are noop when socket_count is 0, but I agree this should return to avoid breakage in the future.

Will change.